pkg/util/events.go

```diff
@@ -0,0 +1,63 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package util
+
+import (
+	"time"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/memorydb"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	svcapitypes "github.com/aws-controllers-k8s/memorydb-controller/apis/v1alpha1"
+)
+
+var (
+	// eventsDuration represents the maximum time range worth of
+	//events to retrieve.
+	eventsDuration = time.Duration(336) * time.Hour
+	// MaxEvents represents the maximum events can to retrieve.
+	MaxEvents = int64(20)
+)
+
+func NewDescribeEventsInput(
+	sourceName *string,
+	sourceType string,
+	maxResults int64,
+) *svcsdk.DescribeEventsInput {
+	input := &svcsdk.DescribeEventsInput{}
+	input.SetSourceType(sourceType)
+	input.SetSourceName(*sourceName)
+	input.SetMaxResults(maxResults)
+	input.SetDuration(int64(eventsDuration.Minutes()))
+	return input
+}
+
+// EventsFromDescribe returns a slice of Event structs given the
+// Output response shape from a DescribeEventsWithContext call
+func EventsFromDescribe(
+	resp *svcsdk.DescribeEventsOutput,
+) []*svcapitypes.Event {
+	var events []*svcapitypes.Event
+	for _, respEvent := range resp.Events {
```

Comment:
Why are you omitting `SourceName` and `SourceType` from this copy operation?

Reply:
I checked online document before and `Event` there didn't have `SourceName` and `SourceType` when I wrote this code. I see `Event` has them in `types.go` file now, and I will add them!
https://pkg.go.dev/github.com/aws-controllers-k8s/memorydb-controller/apis/v1alpha1#Event

Reply:
Added~

Reply:
`SourceName` will always be the name of the `User`/`ACL`/`Snapshot` and `SourceType` will be `"User"`,`"ACL",`"Snapshot"`? This information is already assumed based on the resource that has the events. Do we need to repeat this in every event?

Reply:
@RedbackThomson I used debugger and debugger showed it doesn't auto fill `SourceType` and `SourceName`

Reply:
That's not what I meant. I meant that the values returned by `DescribeEvents` will always have the same `SourceType`. If we called `DescribeEvents` for `ACL` all `SourceType` in the response will be `"ACL"`. When we look at the status of `ACL` it's not useful to have `SourceType`, since we already know the event is for the `ACL` object that we are describing

Reply:
I got you. For me, I think it doesn't matter to have `SourceType` or `SourceName` for `events` in a resource because we can get that information from resource `spec`. However, since we do have these two fields in `events`, so I guess it is better to fill them and not leave them to empty?

Reply:
The reason `SourceType` and `SourceName` exist is that MemoryDB API calls `describe-events` that needs to know which specific resource customer is calling to check events. However, in k8s, customer just calls `kubectl describe resource` and the `events` is a field in the `status` of resource, so `SourceType` and `SourceName` don't impact in this case.

Reply:
Right. If we don't set them in this method, does `sourceType` and `sourceName` show up as empty strings in the status or do they just not appear at all? If they show up as empty strings then I think we should include them, or it will look weird. But if they don't appear, I think it's better to leave them out so that it doesn't clutter the status with useless information

Reply:
Let me test it now

Reply:
If I leave `sourceType` and `sourceName` to empty, they won't show up. Decide to keep them empty.